### PR TITLE
ROX-25498: Allow specifying notifiers by name or ID in policy CRD

### PR DESCRIPTION
**Pull request by clickboo**
### Description

change me!

### User-facing documentation

- [ ] CHANGELOG is updated **OR** update is not needed
- [ ] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

- [ ] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

#### How I validated my change

change me!

### CODE REVIEW

File: config-controller/api/v1alpha1/policy_types.go [hunk 1]
```diff
@@ -154,23 +160,64 @@ func (p SecurityPolicySpec) IsValid() (bool, error) {
...
+			proto.Notifiers = append(proto.Notifiers, notifier)
+			continue
+		}
+		// spec has notifier names specified
+		id, exists := notifiers[notifier]
+		if exists {
+			proto.Notifiers = append(proto.Notifiers, id)
+			continue
+		}
+		log.Warnf("Notifier '%s' does not exist, skipping ..", notifier)
```
[the remark this answers is not in the document]
Comment (kylape) [reply config-controller/api/v1alpha1/policy_types.go:218]: If you agree, may I suggest adding a test that tries to create a policy with an invalid notifier reference and confirm that the reconciliation fails.

File: config-controller/api/v1alpha1/policy_types.go [hunk 2]
```diff
@@ -66,7 +72,7 @@ type SecurityPolicySpec struct {
 	Severity string `json:"severity"`
 	// Enforcement lists the enforcement actions to take when a violation from this policy is identified.  Possible value are UNSET_ENFORCEMENT, SCALE_TO_ZERO_ENFORCEMENT, UNSATISFIABLE_NODE_CONSTRAINT_ENFORCEMENT, KILL_POD_ENFORCEMENT, FAIL_BUILD_ENFORCEMENT, FAIL_KUBE_REQUEST_ENFORCEMENT, FAIL_DEPLOYMENT_CREATE_ENFORCEMENT, and. FAIL_DEPLOYMENT_UPDATE_ENFORCEMENT.
 	EnforcementActions []EnforcementAction `json:"enforcementActions,omitempty"`
-	// Notifiers is a list of IDs of the notifiers that should be triggered when a violation from this policy is identified.  IDs should be in the form of a UUID and are found through the Central API.
+	// Notifiers is a list of IDs or names of the notifiers that should be triggered when a violation from this policy is identified.  IDs should be in the form of a UUID and are found through the Central API.
 	Notifiers []string `json:"notifiers,omitempty"`
```
Comment (kylape) [inline config-controller/api/v1alpha1/policy_types.go:76]: I like making this field dynamically support both UUIDs and names, but have we considered adding a separate field like `notifierNames`?  The current approach matches the UX in the UI but it does diverge from the API.  Just want to make sure we've considered both options.

Comment (clickboo) [reply config-controller/api/v1alpha1/policy_types.go:76]: How does it diverge from the API? The API accepts an array of strings.

Comment (kylape) [reply config-controller/api/v1alpha1/policy_types.go:76]: But if that string is not a UUID, the API rejects it, unlike the behavior in this PR.

Comment (clickboo) [reply config-controller/api/v1alpha1/policy_types.go:78]: Yes, the CRD mimics the UI. Another way of interacting with our API in its current form. I will keep as is, and look into changing the API.

File: config-controller/internal/controller/policy_controller.go [hunk 3]
```diff
@@ -67,13 +67,16 @@ func (r *SecurityPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
 
-	if err := r.PolicyClient.EnsureFresh(ctx); err != nil {
+	if err := r.CentralClient.EnsureFresh(ctx); err != nil {
 		return ctrl.Result{}, errors.Wrap(err, "Failed to refresh")
 	}
 
-	desiredState := policyCR.Spec.ToProtobuf()
+	desiredState, err := policyCR.Spec.ToProtobuf(r.CentralClient.GetNotifiers())
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "Failed to convert policy to protobuf")
```
Comment (kylape) [inline config-controller/internal/controller/policy_controller.go:76]: This now needs to set the status in order to surface invalid notifier name errors to the user via the status field.

**Review by kylape — APPROVED** [review]

**Pull request merged**